### ISSUE TITLE
WorkAction: Add coroutineScope member

### ISSUE THREAD

**Issue by hfhbd** [gradle/gradle#25308]
### Expected Behavior

```kotlin
abstract class UploadAct: WorkAction<WorkParameters.None> {
    override fun execute() {
        coroutineScope.launch {
            upload()
        }
    }
    
    private suspend fun upload() {}
}
```

### Current Behavior (optional)

```kotlin
abstract class UploadAct: WorkAction<WorkParameters.None> {
    override fun execute() = runBlocking {
      upload()
    }
    
    private suspend fun upload() {}
}
```

### Context

I have a Gradle plugin written in Kotlin and I want to upload a file using `WorkAction`. This upload call is asynchronously and the call uses `suspend`, mostly enforced by the http client.
The current method `void execute()` does not support kotlin coroutines and you need to use `runBlocking`, which blocks the thread.
Gradle already supports executing tasks and work action in parallel. Ideally, the WorkQueue acts as a dispatcher and adds coroutineScope to each executing work action to not block the thread, to allow running multiple suspending work action in parallel.

**Comment by eskatos** [comment]
Thank you for your interest in Gradle!

At this stage we won't add coroutines to our Java API. This issue will be closed.

----

You can have the `WorkerExecutor` injected in a worker action to submit more parallel work. See https://docs.gradle.org/current/userguide/custom_gradle_types.html#service_injection

If you want more concurrency control in Worker actions, please open an issue with use cases and requirements that would make it possible to e.g. use coroutines in a worker action.

**Comment by JavierSegoviaCordoba** [comment]
@eskatos it should be great if there was a doc page about how to do that (working with coroutines in the Gradle world).

Is it planned?

**Comment by hfhbd** [comment]
> If you want more concurrency control in Worker actions, please open an issue with use cases and requirements that would make it possible to e.g. use coroutines in a worker action.

What is the difference to this issue?
I understand you don't want to add kotlinx coroutines api to Gradle core, this is the reasons I added the java alternative: `java.util.concurrent.Executor`.

Regarding use-case: I want to upload a file to a http web server. This upload logic is already written and published in a Kotlin jvm library using `suspend` methods. I want to write my workaction in Kotlin to reuse my existing upload library. My current workaround blocks the whole thread of the work action by using `runBlocking`. 

Is it okay to block the whole thread? I (and @JavierSegoviaCordoba) didn't found this in the doc page.

**Comment by eskatos** [comment]
@hfhbd thanks for the feedback. Yes it is ok to block the task execution thread in a task action.

As for the issue management aspect, this issue is specifically geared towards coroutines and adding them. I understand you suggested another approach along the way but please, if you feel strongly about it, open another issue about exposing a managed `Executor`.
